muda/deformers/filter.py

```diff
@@ -0,0 +1,538 @@
+                    duration=obs.duration,
+                    confidence=obs.confidence,
+                    value=new_freq,
+                )
+
+    @staticmethod
+    def filter_midi(annotation, state):
+         #non-existent pitch removed 
+        for obs in annotation.pop_data():
+            new_midi,voice = checkfreqinband(obs.value,state,datatype="midi")
+            if voice:
+                annotation.append(
+                    time=obs.time,
+                    duration=obs.duration,
+                    confidence=obs.confidence,
+                    value=new_midi,
+                )
+
+    @staticmethod
+    def filter_class(annotation, state):
+        #non-existent pitch removed 
+        for obs in annotation.pop_data():
+            value = deepcopy(obs.value)
+            
+            new_freq, voice = checkfreqinband(value["tonic"]+str(value["pitch"]),state,"pitchclass")
+            if voice:
+                value["tonic"] = new_freq[:-1]
+                value["pitch"] = int(new_freq[-1])
+                annotation.append(
+                time=obs.time,
+                duration=obs.duration,
+                confidence=obs.confidence,
+                value=value,
+            )
+            
+           
+
+
+class Filter(AbstractFilter):
+    """ Filtering by cheby2 iir filter
+    
+    This transformation affects the following attributes:
+
+    - Annotations 
+        - pitch_contour, pitch_hz, pitch_midi, pitch_class
+        - note_hz, note_midi
+    - Audio
+
+    Attributes
+    ----------
+    type: "low" or "high" or "bandpass"
+    order: int > 0
+        order of the filter
+    attenutation: float > 0
+        The minimum attenuation required in the stop band. 
+        Specified in decibels, as a positive number.
+
+    cutoff: in hz
+        can be float, list of float, or list of tuples in the case of bandpass filter
+
+    make one or more filters of the same type, but customized cutoff frequencies
+
+    See Also
+    --------
+    RandomFilter
+
+    Examples
+    --------
+    >>> # Shift down by a quarter-tone
+    >>> D = muda.deformers.Filter(btype,order,attenuation,cutoff)
+    """
+
+    def __init__(self,btype="low",order=4, attenuation=60.0, cutoff=4000):
+        AbstractFilter.__init__(self)
+        self.btype = btype
+        self.order = order
+        self.attenuation = attenuation
+        if self.btype == "bandpass":
+            if type(cutoff) == tuple:
+                self.cutoff = [cutoff]
+            elif type(cutoff) == list:
+                if all(type(i) == tuple for i in cutoff):
+                    self.cutoff = cutoff
+                elif all(type(i) == list for i in cutoff):
+                    if all(len(i) == 2 for i in cutoff): # [[a,b],[c,d]]
+                        self.cutoff = [tuple(c) for c in cutoff]
+                    else:
+                        raise ValueError("bandpass filter cutoff must be tuple or list of tuples")
```

Comment:
why not check this first (outside of checking btype) so you don't have multiple copies of this exception?

Reply:
i designed the cutoff variable format to be btype-specific (when btype == bandpass, take list of tuples, tuple or list of length-2 lists; when btype==others, take list of floats or one float), so i can't really check its format outside of knowing what btype it is.. The repetitive exception is each corresponding to a different case: first one is when input is a list of lists of not necessarily length 2, second one is when input is a list of neither tuples nor lists, third one is when input is neither a list nor a tuple. Maybe i should change the name of the exception accordingly to avoid appearing to be repetitive?